lib/whedon/compilers.rb

```diff
@@ -0,0 +1,242 @@
+# This module has methods to compile PDFs and Crossref XML depending upon
+# the content type of the paper (Markdown or LaTeX)
+module Compilers
+
+  # Generate the paper PDF
+  # Optionally pass in a custom branch name as first param
+  def generate_pdf(custom_branch=nil,paper_issue=nil, paper_volume=nil, paper_year=nil)
+    if paper.latex_source?
+      pdf_from_latex(custom_branch, paper_issue, paper_volume, paper_year)
+    elsif paper.markdown_source?
+      pdf_from_markdown(custom_branch, paper_issue, paper_volume, paper_year)
+    end
+  end
+
+  def generate_crossref(paper_issue=nil, paper_volume=nil, paper_year=nil, paper_month=nil, paper_day=nil)
+    if paper.latex_source?
+      crossref_from_latex(paper_issue=nil, paper_volume=nil, paper_year=nil, paper_month=nil, paper_day=nil)
+    elsif paper.markdown_source?
+      crossref_from_markdown(paper_issue=nil, paper_volume=nil, paper_year=nil, paper_month=nil, paper_day=nil)
+    end
+  end
+
+  def pdf_from_latex(custom_branch=nil,paper_issue=nil, paper_volume=nil, paper_year=nil)
+    # Optionally pass a custom branch name
+    `cd #{paper.directory} && git checkout #{custom_branch} --quiet` if custom_branch
+
+    metadata = YAML.load_file("#{paper.directory}/paper.yml")
+
+    for k in ["title", "authors", "affiliations", "keywords", "bibliography"]
+    	raise "Key #{k} not present in metadata" unless metadata.keys().include?(k)
+    end
+
+    # ENV variables or default for issue/volume/year
+    issue = ENV["JLCON_ISSUE"] === nil ? 1 : ENV["JLCON_ISSUE"]
+    volume = ENV["JLCON_VOLUME"] === nil ? 1 : ENV["JLCON_VOLUME"]
+    year = ENV["JLCON_YEAR"] === nil ? 2019 : ENV["JLCON_YEAR"]
+    journal_name = ENV["JOURNAL_NAME"]
+
+    `cd #{paper.directory} && rm -f *.aux \
+    && rm -f *.blg && rm -f *.fls && rm -f *.log\
+    && rm -f *.fdb_latexmk`
+
+    open("#{paper.directory}/header.tex", 'w') do |f|
```

Comment:
Read the metadata from the `paper.yml` file and generate a LaTeX header (`header.tex`)